src/ServerRequestFactoryInterface.php

```diff
@@ -12,20 +12,9 @@
      *
      * @param string $method
      * @param UriInterface|string $uri
+     * @param array $server optional server params ($_SERVER or similar structure)
      *
      * @return ServerRequestInterface
      */
-    public function createServerRequest($method, $uri);
-
-    /**
-     * Create a new server request from server variables.
-     *
-     * @param array $server Typically $_SERVER or similar structure.
-     *
-     * @return ServerRequestInterface
-     *
-     * @throws \InvalidArgumentException
-     *  If no valid method or URI can be determined.
-     */
-    public function createServerRequestFromArray(array $server);
+    public function createServerRequest($method, $uri, $server = []);
```

Comment:
Well, wait a second though.

I took a look to see what Diactoros does, and I think the answer to this one may actually be pretty simple after all.

There is no "preferred source of truth", because in PSR-7, the server-params, URI and method are distinct model values.

So the interface we're defining is for a factory with the domain values already prepared - server-params will be taken and applied exactly as they're given; this should work just fine, and we don't need to define any behavior, but we should specify that there shouldn't *be* any behavior.

This is much better, safer and cleaner than stipulating rules about which server-params must be included and how they're treated, which would be necessary for the currently proposed `createServerRequestFromArray()`, which would need to throw if the given server-params don't include values from which the URI and method can be obtained.

We should stay away from that issue entirely, because we don't need to go near it at all - it's beyond the scope of interoperability.

All we need to define is a means of producing a server-request instance - parsing values from `$_SERVER` is the concern of a *global* request-factory, which you can absolutely build from the single method-signature.

All we need to do is state that `$server` will be taken and applied as given.

There's no reason to go beyond that.